src/source/Ice/IceAgent.c

```diff
@@ -45,7 +45,7 @@ STATUS createIceAgent(PCHAR username, PCHAR password, PIceAgentCallbacks pIceAge
     pIceAgent->localNetworkInterfaceCount = ARRAY_SIZE(pIceAgent->localNetworkInterfaces);
     pIceAgent->candidateGatheringEndTime = INVALID_TIMESTAMP_VALUE;
 
-    pIceAgent->lock = MUTEX_CREATE(FALSE);
+    pIceAgent->lock = MUTEX_CREATE(TRUE);
```

Comment:
@chehefen could you comment on this change?

Reply:
The idea is to not call our apis directly inside our callbacks. Can you set some flag in onIceCandidate and call peerConnectionGetLocalDescription  in your application thread?

Reply:
It this is an open-source library, you should allow API calls from the callbacks, otherwise:
* It forces the application to use another thread, or periodic checks - the unnecessary cost to the app, and I am certainly not interested in going down this route.
* Error-prone, hence the extra cost for you the supporters. (I am sure I wouldn't be the only one attempting to this.)

In another word, a library should avoid forcing a particular design pattern to the users.

So, I'd suggest:
(1) Unlock mutex before callback
(2) Or use recursive/reentrant mutex

I personally think (1) is relatively safer if unlocking and locking again won't cause any atomicity/ordering issues with the function calling the callback.

Any thoughts?

Reply:
@enobufs I would recommend looking deeper into the interlocking issue with @chehefen 

I agree that forcing the customer to spin up a thread might not be the best solution for multitude of reasons. Re-entrant locking might be OK too. Unlocking/re-locking needs to deal with consistency issues after re-locking. 

I will let @chehefen to look into this

Reply:
Great! thanks @MushMal

Reply:
I dont want to make the lock re-entrant because there are places where lock is given up early. If the lock is re-entrant, I dont know if the thread actually given up the lock or not. Unlocking/re-locking needs to deal with consistency issues after re-locking. Ideally we should have an event loop and schedule the callback task there. But since we dont have such construct I would just create a thread in ice to do the callback if you have to make the call within the callback.

Reply:
> I would just create a thread ...
Is the cost justifiable? The cost for that comes from users' budget, you know...

> consistency issues after re-locking
Then, you could arrange the code not to require re-locking, couldn't you? Defer all callbacks and call them after unlocking when they are one-way notifications.

Reply:
can you rebase your PR on top of this commit https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/commit/d9cfcdf0165660969e1eb2cbdee6fabda4dcd9c6? I moved the ice candidate reporting off user thread. You dont need to make the iceAgent lock re-entrant.

Reply:
@chehefen I have reverted those two (reentrant and swap), then rebase on d9cfcdf as you suggested.

I see SDP generated from the new method, but all of those port numbers reported on the SDP are not what the library is actually listening on.

SDP returned by GetLocalDescription():
```
{"type": "offer", "sdp": "v=0\r\no=- 114807987 2 IN IP4 127.0.0.1\r\ns=-\r\nt=0 0\r\na=group:BUNDLE 0\r\na=msid-semantic: WMS myKvsVideoStream\r\nm=application 9 UDP/DTLS/SCTP webrtc-datachannel\r\nc=IN IP4 127.0.0.1\r\na=candidate:3 1 udp 658881 2601:0646:C202:12D0:0000:0000:0000:1B60 43481 typ host raddr ::/0 rport 0 generation 0 network-cost 999\r\na=candidate:2 1 udp 658881 2601:0646:C202:12D0:DD30:417A:2D01:64CC 43225 typ host raddr ::/0 rport 0 generation 0 network-cost 999\r\na=candidate:1 1 udp 658881 2601:0646:C202:12D0:0845:6473:64BB:42AB 42969 typ host raddr ::/0 rport 0 generation 0 network-cost 999\r\na=candidate:0 1 udp 658881 10.0.0.135 42713 typ host raddr 0.0.0.0 rport 0 generation 0 network-cost 999\r\na=rtcp:9 IN IP4 0.0.0.0\r\na=ice-ufrag:Hlv6\r\na=ice-pwd:WOfwKoKbs09L07z41jI7FhGX\r\na=fingerprint:sha-256 4D:FF:37:B5:0D:6D:D4:5F:76:FF:DD:38:FF:E4:DC:8F:75:65:97:C2:E6:E6:AA:A6:00:85:62:C4:74:AC:45:6B\r\na=setup:actpass\r\na=mid:0\r\na=sctp-port:5000\r\n"}
```

What netstat says during the connectivity checking:
```
$ netstat  -an -p udp
Active Internet connections (including servers)
Proto Recv-Q Send-Q  Local Address          Foreign Address        (state)
udp6       0      0  2601:646:c202:12.55721 *.*
udp6       0      0  2601:646:c202:12.55720 *.*
udp6       0      0  2601:646:c202:12.55719 *.*
udp4       0      0  10.0.0.135.55718       *.*
```

All these port numbers on SDP are all swapped as bellow:
```
43481 = 0xa9d9, 0xd9a9 = 55721
43225 = 0xa8d9, 0xd9a8 = 55720
42969 = 0xa7d9, 0xd9a7 = 55719
42713 = 0xa6d9, 0xd9a6 = 55718
```

Reply:
My log printed out what onIceCandidate callback reported also. Same situation. All port numbers are swapped:
```
onIceCandidate: sess=0x7f9fad80cc00 candidate={"candidate":"candidate:3 1 udp 658881 2601:0646:C202:12D0:0000:0000:0000:1B60 43481 typ host raddr ::/0 rport 0 generation 0 network-cost 999","sdpMid":"0","sdpMLineIndex":0}
2020-05-02 20:53:07 DEBUG   iceAgentLogNewCandidate(): New local ice candidate discovered. Id: Lob/Felt4. Ip: 2601:0646:c202:12d0:dd30:417a:2d01:64cc:43225. Type: host
onIceCandidate: sess=0x7f9fad80cc00 candidate={"candidate":"candidate:2 1 udp 658881 2601:0646:C202:12D0:DD30:417A:2D01:64CC 43225 typ host raddr ::/0 rport 0 generation 0 network-cost 999","sdpMid":"0","sdpMLineIndex":0}
2020-05-02 20:53:07 DEBUG   iceAgentLogNewCandidate(): New local ice candidate discovered. Id: 92j/NEDW8. Ip: 2601:0646:c202:12d0:0845:6473:64bb:42ab:42969. Type: host
onIceCandidate: sess=0x7f9fad80cc00 candidate={"candidate":"candidate:1 1 udp 658881 2601:0646:C202:12D0:0845:6473:64BB:42AB 42969 typ host raddr ::/0 rport 0 generation 0 network-cost 999","sdpMid":"0","sdpMLineIndex":0}
2020-05-02 20:53:07 DEBUG   iceAgentLogNewCandidate(): New local ice candidate discovered. Id: MFfUSNlpg. Ip: 10.0.0.135:42713. Type: host
onIceCandidate: sess=0x7f9fad80cc00 candidate={"candidate":"candidate:0 1 udp 658881 10.0.0.135 42713 typ host raddr 0.0.0.0 rport 0 generation 0 network-cost 999","sdpMid":"0","sdpMLineIndex":0}
```

Reply:
@chehefen @MushMal Okay, I found an error in my code. `initKvsWebRtc()` wasn't called. Consequently, getInt16 was using the default being BE.

With the currently pushed branch 64fd2d, my problems have all been resolved.